pilot/pkg/proxy/envoy/v2/internalgen.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"istio.io/istio/pilot/pkg/networking/util"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/golang/protobuf/ptypes/any"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+const (
+	TypeURLConnect = "istio.io/connect"
+	TypeURLDisconnect = "istio.io/disconnect"
+
+	// TODO: TypeURLReady - readiness events for endpoints, agent can propagate
+
+	// TypeURLNACK will receive messages of type DiscoveryRequest, containing
+	// the 'NACK' from envoy on rejected configs. Only ID is set in metadata.
+	// This includes all the info that envoy (client) provides.
+	TypeURLNACK = "istio.io/nack"
+
+)
+
+// InternalGen is a Generator for XDS status updates: connect, disconnect, nacks, acks
+type InternalGen struct {
+	Server *DiscoveryServer
+
+	// TODO: track last N Nacks and connection events, with 'version' based on timestamp.
+	// On new connect, use version to send recent events since last update.
+}
+
+func (sg *InternalGen) OnConnect(node *core.Node) {
+	sg.startPush(TypeURLConnect, []*any.Any{util.MessageToAny(node)})
+}
+
+func (sg *InternalGen) OnDisconnect(node *core.Node) {
+	sg.startPush(TypeURLDisconnect, []*any.Any{util.MessageToAny(node)})
+}
+
+func (sg *InternalGen) OnNack(node *model.Proxy, dr *xdsapi.DiscoveryRequest) {
+	// Make sure we include the ID - the DR may not include metadata
+	dr.Node.Id = node.ID
```

Comment:
We probably don't want to mutate the request (?)

Reply:
I'm not sure - the field was garbage collected before. I can create a copy - and let the old one still get garbage collected. Or document that the meta we hold (and send) has extra info added by the intermediaries.

In a generic 'mesh' model - with MC, etc - we may want to encode more info, the original meta has what the client knows. I'm treating it like the HTTP or mail MIME headers.